src/app/interaction-model/InvokeResponder.h

```diff
@@ -0,0 +1,227 @@
+    ///
+    /// Any error returned by this are final and not retriable
+    /// as a retry for CHIP_ERROR_BUFFER_TOO_SMALL is already built in.
+    CHIP_ERROR Complete(CHIP_ERROR error)
+    {
+        VerifyOrReturnError(!mCompleted, CHIP_ERROR_INCORRECT_STATE);
+        mCompleted     = true;
+        CHIP_ERROR err = mWriter->Complete(error);
+
+        if (err != CHIP_ERROR_BUFFER_TOO_SMALL)
+        {
+            return err;
+        }
+
+        // retry once. Failure to flush is permanent.
+        ReturnErrorOnFailure(mWriter->FlushPendingResponses());
+        return mWriter->Complete(error);
+    }
+
+    /// Sends the specified data structure as a response
+    ///
+    /// This version of the send has built-in RETRY and handles
+    /// Flush/Complete automatically.
+    ///
+    /// Any error returned by this are final and not retriable
+    /// as a retry for CHIP_ERROR_BUFFER_TOO_SMALL is already built in.
+    template <typename ReplyData>
+    CHIP_ERROR Send(const ReplyData & data)
+    {
+        VerifyOrReturnError(!mCompleted, CHIP_ERROR_INCORRECT_STATE);
+        mCompleted     = true;
+        CHIP_ERROR err = data.Encode(ResponseEncoder(ReplyData::GetCommandId()));
+        if (err != CHIP_ERROR_BUFFER_TOO_SMALL)
+        {
+            LogErrorOnFailure(err);
+            err = mWriter->Complete(err);
+        }
+        if (err != CHIP_ERROR_BUFFER_TOO_SMALL)
+        {
+            return err;
+        }
+
+        // retry once. Failure to flush is permanent.
+        ReturnErrorOnFailure(mWriter->FlushPendingResponses());
+        err = data.Encode(ResponseEncoder(ReplyData::GetCommandId()));
+
+        // If encoding fails, we will end up sending an error back to the other side
+        // the caller
+        LogErrorOnFailure(err);
+        if (err == CHIP_NO_ERROR)
+        {
+            err = mWriter->Complete(err);
+        }
+        else
+        {
+            // Error in "complete" is not something we can really forward anymore since
+            // we already got an error in Encode ... just log this.
+            LogErrorOnFailure(mWriter->Complete(err));
+        }
+
+        return err;
+    }
+
+private:
+    InvokeResponder * mWriter;
+    bool mCompleted = false;
+};
+
+enum ReplyAsyncFlags
+{
+    // Some commmands that are expensive to process (e.g. crypto)
+    // should ack right away to improve reliability and reduce
+    // needless retries.
+    kFlushAckRightAway = 0x0001,
+};
+
+class InvokeReply
+{
+public:
+    virtual ~InvokeReply() = default;
+
+    // reply with no data
+    CHIP_ERROR Reply(CHIP_ERROR e) { return this->Reply().Complete(e); }
+
+    // Send a reply "NOW" to the given invoke
+    virtual AutoCompleteInvokeResponder Reply() = 0;
+
+    // Reply "later" to the command. This allows async processing. A reply will be forced
+    // when the returned InvokeReply is destroyed.
+    virtual std::unique_ptr<InvokeReply> ReplyAsync(BitFlags<ReplyAsyncFlags> flags) = 0;
```

Comment:
I see this as a parallel to `CommandHandler::Handle`. With the `CommandHandler::Handle` mechanism, application code had the ability to reply to a command async without a heap. Do we want to attempt to do something similar here instead

Reply:
Now that I have clarified there is multiple InvokeResponse. I think I see the need for `std::unique_ptr`. But still might be good to see if there is a possible way we could do so without allocation.

Reply:
I have added a bit of comments. 

The difficulty I have here is that the calls are virtual since the intent is to be able to unit test these and create mocks. The closest equivalent we have are reference counting where we can use memory pools. I thought that using that path may increase complexity from the start. We can update logic to use ReferenceCounted maybe as a separate work stream.